fsw/cfe-core/src/es/cfe_es_resource.c

```diff
@@ -0,0 +1,428 @@
+        }
+
+       ++CounterRecPtr;
+       --Count;
+    }
+
+    return CounterRecPtr;
+}
+
+
+/*********************************************************************/
+/*
+ * CFE_ES_LocateAppRecordByID
+ *
+ * For complete API information, see prototype in header
+ */
+CFE_ES_AppRecord_t *CFE_ES_LocateAppRecordByID(CFE_ES_ResourceID_t AppID)
+{
+    CFE_ES_AppRecord_t *AppRecPtr;
+    uint32 Idx;
+
+    if (CFE_ES_AppID_ToIndex(AppID, &Idx) == CFE_SUCCESS)
+    {
+        AppRecPtr = &CFE_ES_Global.AppTable[Idx];
+    }
+    else
+    {
+        AppRecPtr = NULL;
+    }
+
+    return AppRecPtr;
+}
+
+/*********************************************************************/
+/*
+ * CFE_ES_LocateLibRecordByID
+ *
+ * For complete API information, see prototype in header
+ */
+CFE_ES_LibRecord_t* CFE_ES_LocateLibRecordByID(CFE_ES_ResourceID_t LibID)
+{
+    CFE_ES_LibRecord_t *LibRecPtr;
+    uint32 Idx;
+
+    if (CFE_ES_LibID_ToIndex(LibID, &Idx) == CFE_SUCCESS)
+    {
+        LibRecPtr = &CFE_ES_Global.LibTable[Idx];
+    }
+    else
+    {
+        LibRecPtr = NULL;
+    }
+
+    return LibRecPtr;
+}
+
+/*********************************************************************/
+/*
+ * CFE_ES_LocateTaskRecordByID
+ *
+ * For complete API information, see prototype in header
+ */
+CFE_ES_TaskRecord_t *CFE_ES_LocateTaskRecordByID(CFE_ES_ResourceID_t TaskID)
+{
+    CFE_ES_TaskRecord_t *TaskRecPtr;
+    uint32 Idx;
+
+    if (CFE_ES_TaskID_ToIndex(TaskID, &Idx) == CFE_SUCCESS)
+    {
+        TaskRecPtr = &CFE_ES_Global.TaskTable[Idx];
+    }
+    else
+    {
+        TaskRecPtr = NULL;
+    }
+
+    return TaskRecPtr;
+}
+
+/*********************************************************************/
+/*
+ * CFE_ES_LocateCounterRecordByID
+ *
+ * For complete API information, see prototype in header
+ */
+CFE_ES_GenCounterRecord_t* CFE_ES_LocateCounterRecordByID(CFE_ES_ResourceID_t CounterID)
+{
+    CFE_ES_GenCounterRecord_t *CounterRecPtr;
+    uint32 Idx;
+
+    if (CFE_ES_CounterID_ToIndex(CounterID, &Idx) == CFE_SUCCESS)
+    {
+        CounterRecPtr = &CFE_ES_Global.CounterTable[Idx];
+    }
+    else
+    {
+        CounterRecPtr = NULL;
+    }
+
+    return CounterRecPtr;
+}
+
+/*********************************************************************/
+/*
+ * CFE_ES_GetTaskRecordByContext
+ *
+ * For complete API information, see prototype in header
+ *
+ * This function does additional validation on the task record
+ * and should only be called when global data is locked.
+ */
+CFE_ES_TaskRecord_t *CFE_ES_GetTaskRecordByContext(void)
+{
+    CFE_ES_TaskRecord_t *TaskRecPtr;
+    CFE_ES_ResourceID_t TaskID;
+
+    /*
+    ** Use the OS task ID to get the ES task record
+    */
+    TaskID = CFE_ES_ResourceID_FromOSAL(OS_TaskGetId());
+    TaskRecPtr = CFE_ES_LocateTaskRecordByID(TaskID);
+
+    /*
+     * Confirm that the entry is actually a match (this requires/assumes
+     * the global data is locked).
+     *
+     * If not a match, return NULL.
```

Comment:
Is there a case where the TaskRecordIsMatch will catch an error? If the TaskRecPtr from line 371 is NULL, then the result will be the same. If the lookup in CFE_ES_LocateTaskRecordByID returns a TaskRecPtr, will the CFE_ES_TaskRecordIsMatch always succeed?

Reply:
Yes -  The "Locate" function just finds the entry in the table where that ID is _supposed_ to be.  It does not check validity of the entry.  This is by design, it is used for all operations - creation, deletion, lookup, etc. and records may or may not be initialized.  The "Locate" function returns NULL only if the resource ID was a completely bad value/wrong type etc.

The "IsMatch" function is the part that actually confirms/checks if the table pointer is actually correct for the intended ID.  It will check for table entries that are uninitialized or may be in use but for a different/non-matching ID - it returns `false` in that case.

So yes - the "Locate" step can succeed but the "IsMatch" can then fail.

Also note that "IsMatch" also needs to be called when locked, but the "Locate" doesn't need a lock, as it doesn't actually validate the entry.